jupyterlab_bokeh_server/server.py

```diff
@@ -141,12 +144,177 @@ def cb():
 
     doc.add_periodic_callback(cb, 200)
 
+try:
+    from pynvml.nvml import *
+
+    nvml_init()
+    ngpus = nvmlDeviceGetCount()
+    gpu_handles = [ nvmlDeviceGetHandleByIndex(i) for i in range(ngpus) ]
+
+    def gpu(doc):
+        fig = figure(title="GPU Usage", sizing_mode="stretch_both", y_range=[0, 100])
+
+        gpu = [ nvmlDeviceGetUtilizationRates( gpu_handles[i] ).gpu+1 for i in range(ngpus) ]
+        left = list(range(len(gpu)))
+        right = [l + 0.8 for l in left]
+        source = ColumnDataSource({"left": left, "right": right, "gpu": gpu})
+        #colors = ["#000003", "#410967", "#932567", "#DC5039"] #, "#FBA40A", "#FCFEA4"]
+        colors = ['#440154', '#404387', '#29788E', '#22A784', '#79D151']
+        mapper = LinearColorMapper(palette=colors, low=0, high=100)
+
+        fig.quad(
+            source=source, left="left", right="right", bottom=0, top="gpu", color={"field": "gpu", "transform": mapper}
+        )
+
+        doc.title = "GPU Utilization [%]"
+        doc.add_root(fig)
+
+        def cb():
+            source.data.update({"gpu": [ nvmlDeviceGetUtilizationRates( gpu_handles[i] ).gpu+0.1 for i in range(ngpus) ]})
+
+        doc.add_periodic_callback(cb, 200)
 
-routes = {
-    "/cpu": cpu,
-    "/resources": resource_timeline,
-}
+    def pci(doc):
+
+        tx_fig = figure(title="TX Bytes [KB]", sizing_mode="stretch_both", y_range=[0, 10000])
+        pci_tx = [ nvmlDeviceGetPcieThroughput( gpu_handles[i] , NVML_PCIE_UTIL_TX_BYTES ) for i in range(ngpus) ]
+        left = list(range(len(pci_tx)))
+        right = [l + 0.8 for l in left]
+        source = ColumnDataSource({"left": left, "right": right, "pci-tx": pci_tx})
+        colors = ["#000003", "#410967", "#932567", "#DC5039"] #, "#FBA40A", "#FCFEA4"]
+        mapper = LinearColorMapper(palette=colors, low=0, high=10000)
```

Comment:
This seems a bit specific to me.  I wonder if there is a way to just have Bokeh decide here.  I would expect to select a color palette like "viridis" and be done with it.  (My expectation could be wrong though.)